mlflow/data/pandas_dataset.py

```diff
@@ -0,0 +1,149 @@
+    Represents a Pandas dataset for use with MLflow Tracking.
+    """
+
+    def __init__(
+        self,
+        df: pd.DataFrame,
+        source: Union[FileSystemDatasetSource, SparkDatasetSource, DeltaDatasetSource],
+        targets: str = None,
+        name: Optional[str] = None,
+        digest: Optional[str] = None,
+    ):
+        """
+        :param df: A pandas dataset.
+        :param source: The source of the pandas dataset.
+        :param targets: The name of the target column. Optional.
+        :param name: The name of the dataset. E.g. "wiki_train". If unspecified, a name is
+                     automatically generated.
+        :param digest: The digest (hash, fingerprint) of the dataset. If unspecified, a digest
+                       is automatically computed.
+        """
+        self._df = df
+        self._targets = targets
+        super().__init__(source=source, name=name, digest=digest)
+
+    def _compute_digest(self) -> str:
+        """
+        Computes a digest for the dataset. Called if the user doesn't supply
+        a digest when constructing the dataset.
+        """
+        MAX_ROWS = 10000
+
+        # drop object columns
+        df = self._df.select_dtypes(exclude=["object"])
+        trimmed_df = df.head(MAX_ROWS)
+        # hash trimmed dataframe contents
+        md5 = hashlib.md5(pd.util.hash_pandas_object(trimmed_df).values)
+        # hash dataframe dimensions
+        n_rows = len(df)
+        md5.update(np.int64(n_rows))
+        # hash column names
+        columns = df.columns
+        for x in columns:
+            md5.update(x.encode())
+        return md5.hexdigest()[:8]
+
+    def _to_dict(self, base_dict: Dict[str, str]) -> Dict[str, str]:
+        """
+        :param base_dict: A string dictionary of base information about the
+                          dataset, including: name, digest, source, and source
+                          type.
+        :return: A string dictionary containing the following fields: name,
+                 digest, source, source type, schema (optional), profile
+                 (optional).
+        """
+        base_dict.update(
+            {
+                "schema": json.dumps({"mlflow_colspec": self.schema.to_dict()}),
+                "profile": json.dumps(self.profile),
+            }
+        )
+        return base_dict
+
+    @property
+    def source(self) -> FileSystemDatasetSource:
+        return self._source
+
+    @property
+    def df(self) -> pd.DataFrame:
+        return self._df
+
+    @property
+    def profile(self) -> Optional[Any]:
+        """
+        A profile of the dataset. May be None if no profile is available.
+        """
+        return {
+            "num_rows": len(self._df),
+            "num_elements": int(self._df.size),
+        }
+
+    @cached_property
+    def schema(self) -> Schema:
+        """
+        An MLflow ColSpec schema representing the columnar dataset
+        """
+        return _infer_schema(self._df)
```

Comment:
Can we try/catch in case schema inference fails and log a warning if it does (see HuggingFace for an example)? Schema inference may fail if there's an unsupported dtype. Can we also change the output to `Optional[Schema]`?